File: devops/ci-cd.js
/**
 * CI/CD - JavaScript
 * 
 * Continuous Integration and Continuous Deployment:
 * - Build pipeline
 * - Testing stages
 * - Deployment automation
 */

// CI/CD Pipeline
class CICDPipeline {
    constructor(name) {
        this.name = name;
        this.stages = [];
        this.currentStage = null;
    }
    
    addStage(stage) {
        this.stages.push(stage);
        return this;
    }
    
    async execute() {
        console.log(`Starting pipeline: ${this.name}`);
        
        for (const stage of this.stages) {
            this.currentStage = stage;
            console.log(`\nExecuting stage: ${stage.name}`);
            
            try {
                await stage.execute();
                console.log(`✓ Stage ${stage.name} completed`);
            } catch (error) {
                console.error(`✗ Stage ${stage.name} failed: ${error.message}`);
                if (stage.required) {
                    throw error;
                }
            }
        }
        
        console.log(`\nPipeline ${this.name} completed successfully`);
    }
}

// Pipeline Stage
class PipelineStage {
    constructor(name, executeFn, required = true) {
        this.name = name;
        this.executeFn = executeFn;
        this.required = required;
    }
    
    async execute() {
        return await this.executeFn();
    }
}

// Build Stage
class BuildStage extends PipelineStage {
    constructor(projectType = 'node') {
        super('Build', async () => {
            console.log(`  Building ${projectType} project...`);
            // Simulate build
            await this.simulateDelay(1000);
            console.log(`  Build artifacts created`);
            return { artifacts: ['dist/', 'build/'] };
        });
    }
    
    simulateDelay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Test Stage
class TestStage extends PipelineStage {
    constructor(testTypes = ['unit', 'integration']) {
        super('Test', async () => {
            console.log(`  Running tests: ${testTypes.join(', ')}`);
            for (const testType of testTypes) {
                console.log(`    Running ${testType} tests...`);
                await this.simulateDelay(500);
                console.log(`    ✓ ${testType} tests passed`);
            }
            return { passed: true, coverage: 85 };
        });
    }
    
    simulateDelay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Deploy Stage
class DeployStage extends PipelineStage {
    constructor(environment = 'staging') {
        super('Deploy', async () => {
            console.log(`  Deploying to ${environment}...`);
            await this.simulateDelay(2000);
            console.log(`  Deployment to ${environment} successful`);
            return { environment, url: `https://${environment}.example.com` };
        });
    }
    
    simulateDelay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Example usage
console.log('=== CI/CD Pipeline ===\n');

const pipeline = new CICDPipeline('Production Deployment');

pipeline
    .addStage(new BuildStage('node'))
    .addStage(new TestStage(['unit', 'integration', 'e2e']))
    .addStage(new DeployStage('staging'))
    .addStage(new DeployStage('production'));

// Execute pipeline (simulated)
pipeline.execute().catch(error => {
    console.error('Pipeline failed:', error);
});

module.exports = { CICDPipeline, PipelineStage, BuildStage, TestStage, DeployStage };

File: devops/ci-cd.py
"""
CI/CD - Python

Continuous Integration and Continuous Deployment:
- Build pipeline
- Testing stages
- Deployment automation
"""

import asyncio
from typing import List, Callable, Dict, Any
from abc import ABC, abstractmethod


# CI/CD Pipeline
class CICDPipeline:
    def __init__(self, name: str):
        self.name = name
        self.stages: List['PipelineStage'] = []
        self.current_stage = None
    
    def add_stage(self, stage: 'PipelineStage'):
        self.stages.append(stage)
        return self
    
    async def execute(self):
        print(f"Starting pipeline: {self.name}")
        
        for stage in self.stages:
            self.current_stage = stage
            print(f"\nExecuting stage: {stage.name}")
            
            try:
                await stage.execute()
                print(f"✓ Stage {stage.name} completed")
            except Exception as error:
                print(f"✗ Stage {stage.name} failed: {error}")
                if stage.required:
                    raise
        
        print(f"\nPipeline {self.name} completed successfully")


# Pipeline Stage
class PipelineStage(ABC):
    def __init__(self, name: str, required: bool = True):
        self.name = name
        self.required = required
    
    @abstractmethod
    async def execute(self) -> Dict[str, Any]:
        pass


# Build Stage
class BuildStage(PipelineStage):
    def __init__(self, project_type: str = 'python'):
        super().__init__('Build')
        self.project_type = project_type
    
    async def execute(self) -> Dict[str, Any]:
        print(f"  Building {self.project_type} project...")
        await asyncio.sleep(1)  # Simulate build
        print("  Build artifacts created")
        return {'artifacts': ['dist/', 'build/']}


# Test Stage
class TestStage(PipelineStage):
    def __init__(self, test_types: List[str] = None):
        super().__init__('Test')
        self.test_types = test_types or ['unit', 'integration']
    
    async def execute(self) -> Dict[str, Any]:
        print(f"  Running tests: {', '.join(self.test_types)}")
        for test_type in self.test_types:
            print(f"    Running {test_type} tests...")
            await asyncio.sleep(0.5)
            print(f"    ✓ {test_type} tests passed")
        return {'passed': True, 'coverage': 85}


# Deploy Stage
class DeployStage(PipelineStage):
    def __init__(self, environment: str = 'staging'):
        super().__init__('Deploy')
        self.environment = environment
    
    async def execute(self) -> Dict[str, Any]:
        print(f"  Deploying to {self.environment}...")
        await asyncio.sleep(2)
        print(f"  Deployment to {self.environment} successful")
        return {'environment': self.environment, 'url': f'https://{self.environment}.example.com'}


# Example usage
async def main():
    print("=== CI/CD Pipeline ===\n")
    
    pipeline = CICDPipeline('Production Deployment')
    
    pipeline \
        .add_stage(BuildStage('python')) \
        .add_stage(TestStage(['unit', 'integration', 'e2e'])) \
        .add_stage(DeployStage('staging')) \
        .add_stage(DeployStage('production'))
    
    # Execute pipeline
    try:
        await pipeline.execute()
    except Exception as error:
        print(f"Pipeline failed: {error}")


if __name__ == "__main__":
    asyncio.run(main())

